lib/Conversion/TorchToLinalg/Random.cpp

```diff
@@ -192,6 +199,301 @@ class ConvertAtenUniformOp : public OpConversionPattern<AtenUniformOp> {
+                                      ValueRange args) {
+      Value jIndex = b.create<arith::IndexCastOp>(loc, indexTy, j);
+
+      Value sum;
+      if (inputRank == 1) {
+        sum = b.create<tensor::ExtractOp>(loc, sumWeights, ValueRange{});
+      } else {
+        sum = b.create<tensor::ExtractOp>(loc, sumWeights, ValueRange{jIndex});
+      }
+
+      // compute cdf in loop
+      Value initCdf = b.create<tensor::EmptyOp>(
+          loc, getAsOpFoldResult(ValueRange{numCategoriesIndex}), elemTy);
+      Value cdf =
+          b.create<scf::ForOp>(
+               loc, cstZero, numCategories, cstOne, ValueRange{initCdf},
+               [&](OpBuilder &b, Location loc, Value i, ValueRange vals) {
+                 Value distribution = vals[0];
+                 // if (i > 0)
+                 auto comparisonPredicate = arith::CmpIPredicateAttr::get(
+                     b.getContext(), arith::CmpIPredicate::sgt);
+                 Value condition = b.create<arith::CmpIOp>(
+                     loc, comparisonPredicate, i, cstZero);
+                 Value iIndex = b.create<arith::IndexCastOp>(loc, indexTy, i);
+                 // curr_cum = i > 0 ? prob[i] + prob[i-1] : prob[i]
+                 ValueRange ind;
+                 if (inputRank == 1) {
+                   ind = ValueRange{iIndex};
+                 } else {
+                   ind = ValueRange{jIndex, iIndex};
+                 }
+                 Value currWeight = b.create<tensor::ExtractOp>(loc, self, ind);
+                 Value currMass = b.create<arith::DivFOp>(loc, currWeight, sum);
+                 Value currCum =
+                     b.create<scf::IfOp>(
+                          loc, condition,
+                          [&](OpBuilder &b, Location loc) {
+                            Value prevI =
+                                b.create<arith::SubIOp>(loc, i, cstOne);
+                            Value prevIndex = b.create<arith::IndexCastOp>(
+                                loc, indexTy, prevI);
+                            Value prevMass = b.create<tensor::ExtractOp>(
+                                loc, distribution, ValueRange{prevIndex});
+                            Value currSum = b.create<arith::AddFOp>(
+                                loc, currMass, prevMass);
+                            b.create<scf::YieldOp>(loc, ValueRange(currSum));
+                          },
+                          [&](OpBuilder &b, Location loc) {
+                            b.create<scf::YieldOp>(loc, ValueRange{currMass});
+                          })
+                         .getResult(0);
+
+                 Value updatedCdf = b.create<tensor::InsertOp>(
+                     loc, currCum, distribution, ValueRange(iIndex));
+                 b.create<scf::YieldOp>(loc, ValueRange(updatedCdf));
+               })
+              .getResult(0);
+
+      // Get key, min and max used by RNG.
+      Value key = b.create<TorchConversion::GetNextSeedOp>(loc);
```

Comment:
`GetNextSeedOp` essentially generates a new global seed for our RNG process. The reason that we recompute this seed at the batch level, but not at the sample level is because the RNG algorithm we are using takes both a key and an incrementing counter as input when generating a sequence of samples - each of which contribute to the randomness of subsequently generated elements. In our case, the key is the random seed from this op and the incrementing counter is the sample index we're currently on in the inner sample loop. 

For more clarity on this, you can see that this is how it is called when lowering the `AtenUniformOp` as well: https://github.com/llvm/torch-mlir/blob/7e7dadf819473e49d5287230ad738abc213cf57f/lib/Conversion/TorchToLinalg/Random.cpp#L156

Also, the RNG algorithm is implemented in this file with a link to the paper: https://github.com/llvm/torch-mlir/blob/7e7dadf819473e49d5287230ad738abc213cf57f/lib/Conversion/TorchToLinalg/Random.cpp#L74

Reply:
Thanks for the info!